document/core/exec/instructions.rst

```diff
@@ -863,101 +824,111 @@ Memory Instructions
 :math:`\MEMORYCOPY`
 ...................
 
-1. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+1. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+
+2. Assert: due to :ref:`validation <valid-memory.copy>`, :math:`F.\AMODULE.\MIMEMS[0]` exists.
+
+3. Let :math:`\X{ma}` be the :ref:`memory address <syntax-memaddr>` :math:`F.\AMODULE.\MIMEMS[0]`.
 
-2. Pop the value :math:`\I32.\CONST~cnt` from the stack.
+4. Assert: due to :ref:`validation <valid-memory.copy>`, :math:`S.\SMEMS[\X{ma}]` exists.
 
-3. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[\X{ma}]`.
 
-4. Pop the value :math:`\I32.\CONST~src` from the stack.
+6. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
 
-5. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+7. Pop the value :math:`\I32.\CONST~n` from the stack.
 
-6. Pop the value :math:`\I32.\CONST~dst` from the stack.
+8. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
 
-7. If :math:`cnt = 0`, then:
+9. Pop the value :math:`\I32.\CONST~s` from the stack.
 
-   a. Return.
+10. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
 
-8. If :math:`cnt = 1`, then:
+11. Pop the value :math:`\I32.\CONST~d` from the stack.
 
-   a. Push the value :math:`\I32.\CONST~dst` to the stack.
+12. If :math:`s + n` is larger than the length of :math:`\X{mem}.\MIDATA` or :math:`d + n` is larger than the length of :math:`\X{mem}.\MIDATA`, then:
 
-   b. Push the value :math:`\I32.\CONST~src` to the stack.
+    a. Trap.
 
-   c. Execute the instruction :math:`\I32\K{.}\LOAD\K{8\_u}~\{ \OFFSET~0, \ALIGN~0 \}`.
+13. If :math:`n = 0`, then:
 
-   d. Execute the instruction :math:`\I32\K{.}\STORE\K{8}~\{ \OFFSET~0, \ALIGN~0 \}`.
+   a. Return.
+
+14. If :math:`d \leq s`, then:
+
+   b. Push the value :math:`\I32.\CONST~d` to the stack.
+
+   c. Execute the instruction :math:`\I32\K{.}\LOAD\K{8\_u}~\{ \OFFSET~0, \ALIGN~0 \}`.
 
-   e. Return.
+   e. Push the value :math:`\I32.\CONST~s` to the stack.
```

Comment:
If I recall correctly, store pops the value first then pops the offset. So the dest offset needs to be pushed before the load pushes the value.

Reply:
Done.